shardingsphere-scaling/shardingsphere-scaling-dialect/shardingsphere-scaling-opengauss/src/main/java/org/apache/shardingsphere/scaling/opengauss/wal/OpenGaussLogicalReplication.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.shardingsphere.scaling.postgresql.wal.event.LogSequenceNumberBase;
+import org.opengauss.PGProperty;
+import org.opengauss.jdbc.PgConnection;
+import org.opengauss.replication.LogSequenceNumber;
+import org.opengauss.replication.PGReplicationStream;
+import org.opengauss.util.PSQLException;
+
+import java.sql.CallableStatement;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Properties;
+
+/**
+ * OpenGauss logical replication.
+ */
+public final class OpenGaussLogicalReplication {
+
+    public static final String SLOT_NAME = "sharding_scaling";
+
+    public static final String DECODE_PLUGIN = "test_decoding";
+
+    public static final String DUPLICATE_OBJECT_ERROR_CODE = "42710";
+
+    /**
+     * Create OpenGauss connection.
+     *
+     * @param jdbcDataSourceConfig JDBC data source configuration
+     * @return OpenGauss connection
+     * @throws SQLException sql exception
+     */
+    public Connection createPgConnection(final StandardJDBCDataSourceConfiguration jdbcDataSourceConfig) throws SQLException {
+        return createConnection(jdbcDataSourceConfig);
+    }
+    
+    private Connection createConnection(final StandardJDBCDataSourceConfiguration jdbcDataSourceConfig) throws SQLException {
+        Properties props = new Properties();
+        PGProperty.USER.set(props, jdbcDataSourceConfig.getHikariConfig().getUsername());
+        PGProperty.PASSWORD.set(props, jdbcDataSourceConfig.getHikariConfig().getPassword());
+        PGProperty.ASSUME_MIN_SERVER_VERSION.set(props, "9.4");
+        PGProperty.REPLICATION.set(props, "database");
+        PGProperty.PREFER_QUERY_MODE.set(props, "simple");
+        return DriverManager.getConnection(jdbcDataSourceConfig.getHikariConfig().getJdbcUrl(), props);
+    }
+    
+    /**
+     * Create OpenGauss replication stream.
+     *
+     * @param pgConnection OpenGauss connection
+     * @param startPosition start position
+     * @return replication stream
+     * @throws SQLException sql exception
+     */
+    public PGReplicationStream createReplicationStream(final PgConnection pgConnection, final LogSequenceNumberBase startPosition) throws SQLException {
+        return pgConnection.getReplicationAPI()
+                .replicationStream()
+                .logical()
+                .withSlotName(SLOT_NAME)
+                .withSlotOption("include-xids", true)
+                .withSlotOption("skip-empty-xacts", true)
+                .withStartPosition((LogSequenceNumber) startPosition.get())
+                .start();
+    }
+
+    /**
+     * Drop exist slots.
+     *
+     * @param conn the datasource connection
+     * @throws SQLException the sql exp
+     */
+    public static void createIfNotExists(final Connection conn) throws SQLException {
+        if (isSlotNameExist(conn)) {
+            dropSlot(conn);
```

Comment:
If slot is dropped here, is replication data lost?

Reply:
no data lost. Before replication slot work, we use inventory task to copy data. So we recreate replication after inventory task is right. avoid og wal_sender thread out of work in previous task.